File: frontend/src/api/colors.ts
import { apiClient } from './client'
import type { Color, ColorCreate } from '../types'

export const colorsApi = {
    /**
     * Get all colors
     */
    getColors: async (): Promise<Color[]> => {
        const response = await apiClient.get<Color[]>('/api/colors/')
        return response.data
    },

    /**
     * Get a single color by ID
     */
    getColor: async (id: string): Promise<Color> => {
        const response = await apiClient.get<Color>(`/api/colors/${id}/`)
        return response.data
    },

    /**
     * Create a new color
     */
    createColor: async (data: ColorCreate): Promise<Color> => {
        const response = await apiClient.post<Color>('/api/colors/', data)
        return response.data
    },

    /**
     * Update an existing color
     */
    updateColor: async (id: string, data: Partial<ColorCreate>): Promise<Color> => {
        const response = await apiClient.put<Color>(`/api/colors/${id}/`, data)
        return response.data
    },

    /**
     * Delete a color
     */
    deleteColor: async (id: string): Promise<void> => {
        await apiClient.delete(`/api/colors/${id}/`)
    },

    /**
     * Search colors by name
     */
    searchByName: async (name: string): Promise<Color[]> => {
        const response = await apiClient.get<Color[]>('/api/colors/', {
            params: { name },
        })
        return response.data
    },
}

export default colorsApi

File: backend/app/api/colors.py
from fastapi import APIRouter, Depends, HTTPException, Query
from typing import List
from app.schemas.color import ColorCreate, ColorResponse
from app.services.color_service import ColorService
from app.core.dependencies import get_color_service

router = APIRouter(prefix="/api/colors", tags=["Colors"])


@router.post("/", response_model=ColorResponse, status_code=201)
async def create_color(
    color_data: ColorCreate,
    service: ColorService = Depends(get_color_service),
):
    """
    Create a new color.

    Returns the created color with its ID.
    """
    try:
        color = service.create_color(color_data.name, color_data.hex_code)
        return color
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")


@router.get("/", response_model=List[ColorResponse])
async def get_colors(
    skip: int = Query(0, ge=0, description="Number of records to skip"),
    limit: int = Query(100, ge=1, le=1000, description="Max records to return"),
    service: ColorService = Depends(get_color_service),
):
    """
    Get all colors with pagination.
    """
    try:
        colors = service.get_all_colors(skip, limit)
        return colors
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")


@router.get("/{color_id}", response_model=ColorResponse)
async def get_color(
    color_id: str,
    service: ColorService = Depends(get_color_service),
):
    """Get a single color by ID"""
    try:
        color = service.get_color_by_id(color_id)
        return color
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")
